X509.pm

```diff
@@ -83,6 +85,103 @@ sub Crypt::OpenSSL::X509::is_selfsigned {
   return $x509->subject eq $x509->issuer;
 }
 
+sub Crypt::OpenSSL::X509::subjectaltname {
+  my $x509 = shift;
+
+  my $ext;
+  eval {
+    # extensions_by_oid croaks of no extensions found
+    # we don't care we will return an empty array
+    $ext = $x509->extensions_by_oid();
+  };
+
+  # Determine whether the SubjectAltName exist
+  if (! defined ${$ext}{'2.5.29.17'}) {
+    # Simply return an empty array if it does not exist
+    my @tmp = ();
+    return \@tmp;
+  }
+
+  my $pdu = ${$ext}{'2.5.29.17'}->value();
+
+  # remove leading '#' from the value returned
+  $pdu =~ s/^#//g;
+
+  my $bin_data = join '', pack 'H*', $pdu;
+  my $asn = Convert::ASN1->new();
+
+  my $ok = $asn->prepare(q<
+    AnotherName ::= SEQUENCE {
+      type    OBJECT IDENTIFIER,
+      value      [0] EXPLICIT ANY } --DEFINED BY type-id }
+
+    EDIPartyName ::= SEQUENCE {
+      nameAssigner            [0]     DirectoryString OPTIONAL,
+      partyName               [1]     DirectoryString }
+
+   --  Directory string type --
+
+    DirectoryString ::= CHOICE {
+      teletexString		TeletexString,  --(SIZE (1..MAX)),
+      printableString		PrintableString,  --(SIZE (1..MAX)),
+      bmpString		BMPString,  --(SIZE (1..MAX)),
+      universalString		UniversalString,  --(SIZE (1..MAX)),
+      utf8String		UTF8String,  --(SIZE (1..MAX)),
+      ia5String		IA5String  --added for EmailAddress
+	}
+
+    AttributeType ::= OBJECT IDENTIFIER
+
+    AttributeValue ::= DirectoryString  --ANY
+
+    AttributeTypeAndValue ::= SEQUENCE {
+      type			AttributeType,
+      value			AttributeValue
+	}
+
+    -- naming data types --
+
+    Name ::= CHOICE { -- only one possibility for now
+      rdnSequence		RDNSequence
+	}
+
+    RDNSequence ::= SEQUENCE OF RelativeDistinguishedName
+
+    DistinguishedName ::= RDNSequence
+
+    RelativeDistinguishedName ::=
+	SET OF AttributeTypeAndValue  --SET SIZE (1 .. MAX) OF
+
+    SubjectAltName ::= GeneralNames
+
+    GeneralNames ::= SEQUENCE OF GeneralName
+
+    GeneralName ::= CHOICE {
+      rfc822Name                      [1]     IA5String,
+      dNSName                         [2]     IA5String,
+      x400Address                     [3]     ANY, --ORAddress,
+      directoryName                   [4]     Name,
+      ediPartyName                    [5]     EDIPartyName,
+      uniformResourceIdentifier       [6]     IA5String,
+      iPAddress                       [7]     OCTET STRING,
+      registeredID                    [8]     OBJECT IDENTIFIER
+    }
+
+  >);
+  die "*** Could not prepare definition: ".$asn->error()
```

Comment:
Please use single-quotes for literals as suggested by [Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals](https://metacpan.org/pod/Perl::Critic::Policy::ValuesAndExpressions::ProhibitInterpolationOfLiterals)

```suggestion
  die '*** Could not prepare definition: '.$asn->error()
```